src/Data/Foldable/Linear.hs

```diff
@@ -0,0 +1,177 @@
+  , all
+  , and
+  , or
+  , traverse_
+  , sequence_
+  , foldMapDefault
+  ) where
+
+import qualified Prelude as Prelude
+import Prelude (Maybe(..), Either(..), Int, Ordering (..))
+import Prelude.Linear.Internal.Simple
+import Data.Unrestricted.Linear
+import Data.Eq.Linear
+import Data.Ord.Linear
+import Data.Bool.Linear
+import Data.Num.Linear
+import Data.Maybe.Linear
+import Data.List.NonEmpty (NonEmpty (..))
+import Data.Functor.Linear
+import Data.Monoid.Linear
+
+class Foldable t where
+    {-# MINIMAL foldMap | foldr #-}
+
+    fold :: Monoid m => t m #-> m
+    fold = foldMap id
+
+    foldMap' :: Monoid m => (a #-> m) -> t a #-> m
+    foldMap' f = foldl' (\acc a -> acc <> f a) mempty
+
+    foldMap :: Monoid m => (a #-> m) -> t a #-> m
+    {-# INLINE foldMap #-}
+    foldMap f = foldr ((<>) . f) mempty
+
+    foldr :: (a #-> b #-> b) -> b #-> t a #-> b
+    foldr f z t = appEndo (foldMap (Endo . f) t) z
+
+    foldr' :: forall a b. (a #-> b #-> b) -> b #-> t a -> b
+    foldr' f z0 xs = foldl f' id xs z0
+      where
+        f' :: (b #-> b) #-> a #-> b #-> b
+        f' k x z = k $! f x z
+
+    foldl :: (b #-> a #-> b) -> b #-> t a #-> b
+    foldl f z t = appEndo (getDual' (foldMap (Dual . Endo . flip f) t)) z
+      where
+        getDual' :: Dual a #-> a
+        getDual' (Dual a) = a
+
+    foldl' :: forall a b. (b #-> a #-> b) -> b #-> t a #-> b
+    foldl' f z0 xs = foldr f' id xs z0
+       where
+         f' :: a #-> (b #-> b) #-> b #-> b
+         f' x k z = k $! f z x
+
+    foldr1 :: forall a. (a #-> a #-> a) -> t a #-> a
+    foldr1 f xs = fromMaybe (Prelude.errorWithoutStackTrace "foldr1: empty structure")
+                    (foldr mf Nothing xs)
+      where
+        mf :: a #-> Maybe a #-> Maybe a
+        mf x Nothing = Just x
+        mf x (Just y) = Just (f x y)
+
+    foldl1 :: forall a. (a #-> a #-> a) -> t a #-> a
+    foldl1 f xs = fromMaybe (Prelude.errorWithoutStackTrace "foldl1: empty structure")
+                    (foldl mf Nothing xs)
+      where
+        mf :: Maybe a #-> a #-> Maybe a
+        mf Nothing y = Just y
+        mf (Just x) y = Just (f x y)
+
+    toList :: t a #-> [a]
+    toList = foldr (:) []
+
+    -- | This is likely not very useful and only here to preserve the symmetry
+    -- between 'Prelude.Foldable'.
+    null :: Consumable a => t a #-> Bool
+    null = foldr (\x a -> (x, a) `lseq` False) True
+
+    length :: Consumable a => t a #-> Int
+    length = foldl' (\c a -> a `lseq` c + 1) 0
+
+    -- | This is likely not very useful and only here to preserve the symmetry
+    -- between 'Prelude.Foldable'.
+    elem :: Prelude.Eq a => a -> t a #-> Bool
+    elem x = any (== x)
+
+    minimum :: (Dupable a, Prelude.Ord a) => t a #-> a
+    minimum = minimumBy compare
+
+    maximum :: (Dupable a, Prelude.Ord a) => t a #-> a
+    maximum = maximumBy compare
+
+instance {-# OVERLAPPABLE #-} (Foldable t, Consumable a) => Consumable (t a) where
+  consume = foldl' (flip lseq) ()
```

Comment:
Do you think this orphan instance is a good idea? I just observed that it causes confusing error messages when something is missing a `Consumable` instance, where you get something along the lines of `X is not a Foldable`. I think we can expose a function or a newtype wrapper instead.

Reply:
No. It's probably not a good idea. I hadn't noticed it before.

Better is to make a deriving-via combinator.